deeppavlov/core/trainers/nn_trainer.py

```diff
@@ -159,15 +159,20 @@ def _validate(self, iterator: DataLearningIterator,
             self.tb_valid_writer.flush()
 
         m_name, score = metrics[0]
-        if self.improved(score):
+        if self.best is not None:
+            if self.improved(score):
+                self.patience = 0
+                log.info('New best {} of {}'.format(m_name, score))
+                self.best = score
+                log.info('Saving model')
+                self.save()
+            else:
+                self.patience += 1
+                log.info('Did not improve on the {} of {}'.format(m_name, self.best))
+        else:
             self.patience = 0
-            log.info('New best {} of {}'.format(m_name, score))
+            log.info('Initial best {} of {}'.format(m_name, score))
             self.best = score
-            log.info('Saving model')
-            self.save()
```

Comment:
So if the first validation gives the best metrics then the model will be saved after training with the worse score because `if not self._saved` will activate.
Also, if `self.validate_first` is `False` then after initial validation the model should also be saved.

Reply:
> So if the first validation gives the best metrics then the model will be saved after training with the worse score 
I think we can't call result 'best' before at least 2 measurements. My current intent is to save model at exit before we clearly see an improvement.

> Also, if self.validate_first is False then after initial validation the model should also be saved.
OK, please help me to formulate the rules of desired behavior.

For the validation savepoint:
1) Don't save models during 'validate_first=True' validation (lets call it 0-th validation)
2) Always save the model during 1st validation (to capture early training results).
3) Starting from the 2nd validation, only save the model if the target metric was improved

For the at-train-exit savepont:
3) Save the model if it happened before 1st validation (which is not 'validate_first'), don't save otherwise.

Is it correct?

Reply:
If we're not improving after 0-th validation, we should not save

Reply:
So the rules shold become like follows, please review.

For the validation savepoint:
* 0-th validation (optional). Don't save model, establish a baseline.
* 1-th validation.
   + If we have a baseline and see an improvement, save the model, don't save otherwise.
   + If we don't have a baseline, save the model.
* 2nd and later validations. Save the model if we see an improvement

For the at-train-exit savepont:
* Save the model if it happened before 1st validation (to capture early training results), don't save otherwise.

Reply:
Sounds ok, yes.
But we need to remember to not save at the end if all validations after 0-th did not improve.

Reply:
> But we need to remember to not save at the end if all validations after 0-th did not improve.

Yes, I see, OK, I'll try to implement it.